pkg/codegen/hcl2/rewrite_convert.go

```diff
@@ -116,5 +118,87 @@ func RewriteConversions(x model.Expression, to model.Type) model.Expression {
 	if to.AssignableFrom(x.Type()) && sameSchemaTypes(to, x.Type()) {
 		return x
 	}
+	// If we can convert a primitive value in place, do so.
+	if value, ok := convertPrimitiveValues(x, to); ok {
+		return value
+	}
 	return NewConvertCall(x, to)
 }
+
+// convertPrimitiveValues returns a new expression if the given expression can be converted to another primitive type
+// (bool, int, number, string) that matches the target type.
+func convertPrimitiveValues(from model.Expression, to model.Type) (model.Expression, bool) {
+	var expression model.Expression
+	switch {
+	case to.AssignableFrom(model.DynamicType):
+		return nil, false
+	case to.AssignableFrom(model.BoolType):
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if b, err := strconv.ParseBool(stringLiteral); err == nil {
```

Comment:
I realize now that in my proposed code I forgot to indicate where the `convert.Convert` calls were coming from. These calls are provided by `github.com/zclconf/go-cty/cty/convert`, and are exactly the APIs used when evaluating HCL2 expressions at runtime. We should use them here as well for correctness.

Reply:
Ah, I see now. Changed to `convert.Convert`.